### PR TITLE
👷 A new "publisher" generic GitHub Action is added

### PR DESCRIPTION
**Pull request by parfeniukink**
### SUMMARY:
* `neuralmagic/nm-actions/actions/publish_pypi` could be used by other GitHub actions to perform builds and publish to distribution indexes
* All the input parameters of the action are optional
* By default this GitHub action script builds and publishes the distribution to the internal packages index
  * If you need to publish the package to the pypi.org, the `publish_pypi` must be specified
  * Some other useful input parameters are added. For more information have a look at the README.md file
  * Now the `publish_pypi` action uses the `utils/inject_build_props.py` script so you can change the build process using its configurations

### Usage example 1:
```yaml
steps:
  - name:
    uses: neuralmagic/nm-actions/actions/publish_pypi@main
    with:
      publish_pypi: true
      publish_pypi_internal: true
```
### Usage example 2:
```yaml
steps:
  - name:
    uses: neuralmagic/nm-actions/actions/publish_pypi@main
    with:
      publish_pypi: true
      publish_pypi_internal: true
```

### CODE REVIEW

File: actions/publisher/action.yml [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+  publish_pypi:
+    description: "Publish the distribution to the pypi.org"
+    required: true
+
+  build_number:
+    description: "Typically the PR number or the Run ID as a fallback solution"
+    required: false
+    default: ${{ github.event.pull_request.number || github.run_id }}
+
+  prefix:
```
Comment (markurtz) [inline actions/publisher/action.yml]: Since we're adding it after the name, should be suffix.

File: actions/publisher/action.yml [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    description: "Determines whether the prefix will added to the distribution name or not"
+    required: false
+    default: false
+
+  timestamp:
```
Comment (markurtz) [inline actions/publisher/action.yml]: Can we combine this with build_number? Specifically, can we change it out for something like an enum with an override choice? So, the user can select the build number that should be applied to change extend the version, or none at all:
- None
- timestamp
- github
- {TEXT}

**Review by markurtz — CHANGES_REQUESTED** [review]
Left a few comments inline. Some additional notes:
- Can we update the name to include pypi or python in the name 
- For the version, we need to edit the toml file before running the build, otherwise the version isn't propagated to the internal code for the library such as guildellm.__version__
- For the name, I believe it's the same requirement, where the toml should be updated so that the change propagates for the library name (we still want to keep source the same so imports are the same)

File: actions/publisher/action.yml (hunk 1 above)
Comment (parfeniukink) [reply actions/publisher/action.yml]: oh.. true :)

File: actions/publish_pypi/action.yml [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+inputs:
+  publish_pypi_internal:
+    description: "Publish the distribution to the internal PyPI server"
+    required: true
+
+  publish_pypi:
+    description: "Publish the distribution to the pypi.org"
+    required: true
+
+  built_type:
```
Comment (markurtz) [inline actions/publish_pypi/action.yml]:
```suggestion
  build_type:
```

File: actions/publish_pypi/action.yml [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+            echo "Invalid build_type: ${{ inputs.built_type }}"
+            exit 1
+        esac
+
+        echo "::set-output name=suffix::$SUFFIX"
+
+    - name: Rename build artifacts
+      run: |
+        # Retrieve package name from previous step
+        PACKAGE_NAME="${{ inputs.custom_package_name || 'guidellm' }}"
```
Comment (markurtz) [inline actions/publish_pypi/action.yml]: Reset this so we are only pulling from custom_package_name and no defaults

File: actions/publish_pypi/action.yml [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+          exit 1
+        fi
+
+    - name: Install tox
+      run: python3 -m pip install --user tox build
+
+    - name: Build the distribution with tox
+      id: build
+      run: |
+        python3 -m tox -e build
```
Comment (markurtz) [inline actions/publish_pypi/action.yml]: Before this line, run toml edits for name and version: https://github.com/neuralmagic/guidellm/blob/main/utils/inject_build_props.py#L46C9-L47C45

File: actions/publish_pypi/action.yml [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+
+  publish_pypi:
+    description: "Publish the distribution to the pypi.org"
+    required: true
+
+  built_type:
+    description: "Applies a set of rules depending on the environment. Available values: (dev|release|nightly|custom)"
+    required: true
+
+  custom_package_name:
```
Comment (markurtz) [inline actions/publish_pypi/action.yml]: Either need to set to package_name and always expect from the user to override in the toml or we should auto load the current package name from the toml and append the suffixes with no defaulting of the package names to guidellm or anything else

**Review by markurtz — CHANGES_REQUESTED** [review]

File: actions/publish_pypi/action.yml (hunk 5 above)
Comment (parfeniukink) [reply actions/publish_pypi/action.yml]: @markurtz I just realized that the tox `build` command already runs `python utils/inject_build_props.py` which is a part of a concrete project. Like, do we still need it right here? In that case we limit all the applications to have this script in that exact folder, isn't it?

File: actions/publish_pypi/action.yml (hunk 6 above)
Comment (parfeniukink) [reply actions/publish_pypi/action.yml]: Just could be removed because of the `inject_build_props.py`

**Review by markurtz — APPROVED** [review]

**Pull request merged**